src/main/kotlin/com/ekino/oss/gradle/plugin/quality/QualityPlugin.kt

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2019 ekino (https://www.ekino.com/)
+ */
+
+package com.ekino.oss.gradle.plugin.quality
+
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.plugins.JavaPlugin
+import org.gradle.api.plugins.quality.Checkstyle
+import org.gradle.api.plugins.quality.CheckstyleExtension
+import org.gradle.api.plugins.quality.CheckstylePlugin
+import org.gradle.api.resources.TextResource
+import org.gradle.api.tasks.SourceSetContainer
+import org.gradle.kotlin.dsl.*
+import org.gradle.testing.jacoco.plugins.JacocoPlugin
+import org.gradle.testing.jacoco.plugins.JacocoPluginExtension
+import org.gradle.testing.jacoco.tasks.JacocoReport
+import org.sonarqube.gradle.SonarQubeExtension
+import org.sonarqube.gradle.SonarQubePlugin
+import org.sonarqube.gradle.SonarQubeTask
+
+class QualityPlugin: Plugin<Project> {
+
+  override fun apply(project: Project) {
+    with(project) {
+
+      // plugins
+      apply<JavaPlugin>()
+      apply<CheckstylePlugin>()
+      apply<JacocoPlugin>()
+      apply<SonarQubePlugin>()
+
+      val sourceSets = the<SourceSetContainer>()
+
+      // configuration
+      configure(listOf(project)) {
+
+        // Checkstyle configuration
+        configure<CheckstyleExtension> {
+          toolVersion = "8.19"
+        }
+
+        tasks.withType(Checkstyle::class.java) {
+          config = getTextResource(project, "checkstyle.xml")
+        }
+
+        // Jacoco configuration
+        configure<JacocoPluginExtension> {
+          toolVersion = "0.8.3"
+        }
+
+        val jacocoTestReport by tasks.named<JacocoReport>("jacocoTestReport") {
+          doFirst {
+            println("Generating jacoco coverage report in HTML ...")
+          }
+          // To add ".exec" file from integrationTest task if exists
+          tasks.findByName("integrationTest")?.let {
+            executionData.from(file("${buildDir}/jacoco/integrationTest.exec")) // To merge Test Tasks Jacoco reports
+          }
+          if (project.hasProperty("sonarCoverageExclusions")) {
+            val excludes = (findProperty("sonarCoverageExclusions") as String).replace(".java", ".class").split(",\\s*".toRegex())
+            classDirectories.setFrom(sourceSets["main"].output.asFileTree.matching {
```

Comment:
There is no special reason for loosing `afterEvaluate`. I iterate many times to make the code work. But it seems this code is executed lazily. I will try to reintegrate the `afterEvaluate` and see if it works.
The `.replace(".java", ".class")` is because Jacoco `classDirectories` refers to `.class` files. And Sonarqube talk about `.java` source files. We use the same property for both.
I still don't know how it has worked before with `.java` patterns for Jacoco.

Reply:
Done (afterEvaluate reintegrated !)

Reply:
ok thanks for the explanation, never had the issue before (probably because we've never filtered with the file extension 😛). I think you can use `sourceSets["main"].output.classesDirs` to reduce the number of files to analyze, since `output` also includes the resources

Reply:
You're absolutely right ! Done !